jxl/src/features/spline.rs

```diff
@@ -270,11 +347,213 @@ pub struct Splines {
     segment_y_start: Vec<usize>,
 }
 
+struct ExtendedPoints<'a> {
+    first_element: Point,
+    last_element: Point,
+    backend: &'a Vec<Point>,
+}
+
+impl ExtendedPoints<'_> {
+    fn len(&self) -> usize {
+        self.backend.len() + 2
+    }
+    fn iter(&self) -> ExtendedPointsIterator {
+        ExtendedPointsIterator {
+            index: 0,
+            backend: self,
+        }
+    }
+}
+
+impl Index<usize> for ExtendedPoints<'_> {
+    type Output = Point;
+    fn index(&self, index: usize) -> &Self::Output {
+        if index == 0 {
+            &self.first_element
+        } else if index == self.len() - 1 {
+            &self.last_element
+        } else {
+            &self.backend[index - 1]
+        }
+    }
+}
+
+struct ExtendedPointsIterator<'a> {
+    index: usize,
+    backend: &'a ExtendedPoints<'a>,
+}
+
+impl Iterator for ExtendedPointsIterator<'_> {
+    type Item = Point;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index < self.backend.len() {
+            let result = Some(self.backend[self.index]);
+            self.index += 1;
+            result
+        } else {
+            None
+        }
+    }
+}
+
+fn draw_centripetal_catmull_rom_spline(points: &Vec<Point>) -> Vec<Point> {
+    if points.is_empty() {
+        return [].to_vec();
+    }
+    if points.len() == 1 {
+        return vec![points[0]];
+    }
+    const NUM_POINTS: usize = 16;
+    // Create a view of points with one prepended and one appended point.
+    let extended_points = ExtendedPoints {
+        first_element: points[0] + (points[0] - points[1]),
+        last_element: points[points.len() - 1]
+            + (points[points.len() - 1] - points[points.len() - 2]),
+        backend: points,
+    };
+    // Precompute the deltas used in the loop.
+    let deltas: Vec<f32> = extended_points
+        .iter()
+        .enumerate()
+        .map(|(index, point)| {
+            if index + 1 < extended_points.len() {
+                (extended_points[index + 1] - point).abs().sqrt()
+            } else {
+                0.0
+            }
+        })
+        .collect();
+    let mut result = Vec::<Point>::with_capacity((points.len() - 1) * NUM_POINTS + 1);
```

Comment:
please use the try_reserve() methods :-)

Reply:
Done